heaps/min_heap.py

```diff
@@ -57,20 +67,51 @@ def heap_up(self, index):
             property is reestablished.
             
             This could be **very** helpful for the add method.
-            Time complexity: ?
-            Space complexity: ?
+            Time complexity: ? O(log n)
+            Space complexity: ? O(1)
         """
-        pass
+        parent_index = int((index - 1) / 2)
+        current_key = self.store[index].key 
+        parent_key = self.store[parent_index].key 
 
+        while current_key < parent_key:
+            self.swap(index, parent_index)
+            index = parent_index
+            parent_index = int((index - 1) / 2)
+            current_key = self.store[index].key
+            parent_key = self.store[parent_index].key 
+
+    def has_left_child(self, index):
+        left_child_index = int((2 * index) + 1)
+        if left_child_index <= len(self.store) - 1:
+            return True
+        return False
+
+    def has_right_child(self, index):
+        right_child_index = int((2 * index) + 2)
+        if right_child_index <= len(self.store) - 1:
+            return True
+        return False
+    
     def heap_down(self, index):
         """ This helper method takes an index and 
             moves the corresponding element down the heap if it's 
             larger than either of its children and continues until
             the heap property is reestablished.
+            time complexity O(log n)
+            space complexity O(1)
```

Comment:
👀 Nice of you to include the complexity here when we forgot to ask for it!

This is where the O(log n) time and space complexity in `remove` comes from. Since `heap_down` calls itself recursively, the space complexity will grow with the depth of the stack, which is the height of the heap, or O(log n).